jax/image/patches.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Sequence, Union, Tuple, Any
+import string
+import numpy as np
+from jax import lax
+from jax import numpy as jnp
+
+
+Array = Any
+PrecisionType = Any
+
+
+def extract_image_patches(
+    lhs: Array,
+    filter_shape: Sequence[int],
+    window_strides: Sequence[int],
+    padding: Union[str, Sequence[Tuple[int, int]]],
+    lhs_dilation: Sequence[int] = None,
+    rhs_dilation: Sequence[int] = None,
+    dimension_numbers: Tuple[str, str, str] = None,
+    precision: PrecisionType = None,
+) -> Array:
+  """Extract patches subject to the receptive field of a general convolution.
+
+  Runs the input through a convolution with given parameters. The kernel of the
+  convolution is constructed such that the output channel dimension `"C"` is
+  replaced with image patches, so instead of a single `"C"` dimension, there is,
+  for example, three dimensions `"chw". The exact order of these dimensions is
+  `"c" + ''.join(c for c in rhs_spec if c not in 'OI')`, where
+  `rhs_spec == dimension_numbers[1]`.
+
+  Docstring below adapted from `jax.lax.conv_general_dilated`.
+
+  See Also:
+    https://www.tensorflow.org/xla/operation_semantics#conv_convolution
+
+  Args:
+    lhs: a rank `n+2` dimensional input array.
+    filter_shape: a sequence of `n` integers, representing the receptive window
+      spatial shape in the order as specified in
+      `rhs_spec = dimension_numbers[1]`.
+    window_strides: a sequence of `n` integers, representing the inter-window
+      strides.
+    padding: either the string `'SAME'`, the string `'VALID'`, or a sequence of
+      `n` `(low, high)` integer pairs that give the padding to apply before and
+      after each spatial dimension.
+    lhs_dilation: `None`, or a sequence of `n` integers, giving the
+      dilation factor to apply in each spatial dimension of `lhs`. LHS dilation
+      is also known as transposed convolution.
+    rhs_dilation: `None`, or a sequence of `n` integers, giving the
+      dilation factor to apply in each spatial dimension of `rhs`. RHS dilation
+      is also known as atrous convolution.
+    dimension_numbers: either `None`, or a 3-tuple
+      `(lhs_spec, rhs_spec, out_spec)`, where each element is a string
+      of length `n+2`. `None` defaults to `("NCHWD..., OIHWD..., NCHWD...")`.
```

Comment:
I find all these references to "LHS" and "RHS" a little confusing. What do they mean in the context of extracting image patches from a single image?

Reply:
These mean the batch of inputs and the (imaginary) convolutional kernel - please see my other comment on trying to follow the `lax.conv_general_dilated` API.